src/parcsr_ls/par_mgr.c

```diff
@@ -3840,166 +4181,466 @@ HYPRE_Int hypre_blockRelax_solve (hypre_ParCSRMatrix *A,
-                          HYPRE_Real          *diaginv,
-                          hypre_ParVector    *Vtemp)
+HYPRE_Int
+hypre_BlockDiagInvLapack(HYPRE_Real *diag, HYPRE_Int N, HYPRE_Int blk_size)
+{
+   HYPRE_Int nblock, left_size, i;
+   //HYPRE_Int *IPIV = hypre_CTAlloc(HYPRE_Int, blk_size, HYPRE_MEMORY_HOST);
+   HYPRE_Int LWORK = blk_size * blk_size;
+   HYPRE_Real *WORK = hypre_CTAlloc(HYPRE_Real, LWORK, HYPRE_MEMORY_HOST);
+   HYPRE_Int INFO;
+
+   HYPRE_Real wall_time;
+   HYPRE_Int my_id;
+   MPI_Comm comm = hypre_MPI_COMM_WORLD;
+   hypre_MPI_Comm_rank(comm, &my_id);
+
+   nblock = N / blk_size;
+   left_size = N - blk_size * nblock;
+   HYPRE_Int *IPIV = hypre_CTAlloc(HYPRE_Int, blk_size, HYPRE_MEMORY_HOST);
+
+   wall_time = time_getWallclockSeconds();
+   if (blk_size >= 2 && blk_size <= 4)
+   {
+      for (i = 0; i < nblock; i++)
+      {
+         hypre_MGRSmallBlkInverse(diag + i * LWORK, blk_size);
+         //hypre_blas_smat_inv_n2(diag+i*LWORK);
+      }
+   }
+   else if (blk_size > 4)
+   {
+      for (i = 0; i < nblock; i++)
+      {
+         hypre_dgetrf(&blk_size, &blk_size, diag + i * LWORK, &blk_size, IPIV, &INFO);
+         hypre_dgetri(&blk_size, diag + i * LWORK, &blk_size, IPIV, WORK, &LWORK, &INFO);
+      }
+   }
+
+   // Left size
+   if (left_size > 0)
+   {
+      hypre_dgetrf(&left_size, &left_size, diag + i * LWORK, &left_size, IPIV, &INFO);
+      hypre_dgetri(&left_size, diag + i * LWORK, &left_size, IPIV, WORK, &LWORK, &INFO);
+   }
+   wall_time = time_getWallclockSeconds() - wall_time;
+   //if (my_id == 0) hypre_printf("Proc = %d, Compute inverse time: %1.5f\n", my_id, wall_time);
+
+   hypre_TFree(IPIV, HYPRE_MEMORY_HOST);
+   hypre_TFree(WORK, HYPRE_MEMORY_HOST);
+
+   return hypre_error_flag;
+}
+// Extract the block diagonal part of a A or a principal submatrix of A defined by a marker (point_type)
+// in an associated CF_marker array. The result is an array of (flattened) block diagonals.
+// If CF marker array is NULL, it returns an array of the (flattened)
+// block diagonal of the entire matrix A. Options for diag_type are:
+// diag_type = 1: return the inverse of the block diagonals
+// otherwise : return the block diagonals
+// On return, blk_diag_size contains the size of the returned (flattened) array.
+// (i.e. nnz of extracted block diagonal)
+// ***Input***
+// A - parCSR matrix
+// blk_size - Size of diagonal blocks to extract
+// CF_marker - Array prescribing submatrix from which to extract block diagonals. Ignored if NULL.
+// point_type - marker tag in CF_marker array to extract diagonal
+// diag_type - Type of block diagonal entries to return. Currently supports block diagonal or inverse block diagonal entries (diag_type = 1).
+// ***Output***
+// diag_ptr: Array of block diagonal entries
+// blk_diag_size - number of entries in extracted block diagonal (size of diag_ptr).
+//
+HYPRE_Int
+hypre_ParCSRMatrixExtractBlockDiag(hypre_ParCSRMatrix   *A,
```

Comment:
Since this operates on `hypre_ParCSRMatrix`, should we move this function definition to some file in the `parcsr_mv` directory? e.g., `par_csr_matrix.c`

Reply:
Noted. I will be refactored to use hypre_ParcsrBdiagInvScal later.